llms/mlx_lm/models/recurrent_gemma.py

```diff
@@ -0,0 +1,505 @@
+import math
+from dataclasses import dataclass
+from typing import List, Literal, Optional
+
+import mlx.core as mx
+import mlx.nn as nn
+
+from .base import BaseModelArgs
+
+
+@dataclass
+class ModelArgs(BaseModelArgs):
+    model_type: str
+    hidden_size: int
+    attention_bias: bool
+    conv1d_width: int
+    embeddings_scale_by_sqrt_dim: bool
+    hidden_size: int
+    intermediate_size: int
+    logits_soft_cap: float
+    num_attention_heads: int
+    num_hidden_layers: int
+    num_key_value_heads: int
+    rms_norm_eps: float
+    rope_theta: float
+    attention_window_size: int
+    vocab_size: int
+    _block_types: List[str]
+
+
+def create_window_causal_mask(N: int, window_size: int):
+    inds = mx.arange(N)
+    linds = inds[:, None]
+    rinds = inds[None]
+    mask = (linds < rinds) | (linds > rinds + window_size)
+    return mask * -1e9
+
+
+class RecurrentCache:
+
+    def __init__(self):
+        self._cache = (None, None)
+
+    def __getitem__(self, idx):
+        return self._cache[idx]
+
+    def update(self, conv_state, recurrent_state):
+        self._cache = (conv_state, recurrent_state)
+
+
+class WindowKVCache:
+
+    def __init__(self, window_size):
+        self.keys = None
+        self.values = None
+        self.offset = 0
+        self.window_size = window_size
+
+    def update_and_fetch(self, keys, values):
+        # TODO consider using rotating buffer here
+        # especially for very long generations
+        def _update(x, v):
+            t = x.shape[2] - self.window_size
```

Comment:
Should that also have `- v.shape[2]` ie `x.shape[2] - self.window_size - v.shape[2]` such that the returned value is of size `self.window_size` at axis 2?

Reply:
A couple things, the context size is actually `window_size + 1`. That's why there is no `+1` in there. Also `v.shape[2]` is always 1 so it doesn't matter too much. But if `v.shape[2] > 1` then my understanding is we don't want to truncate the cache which would drop some tokens that we are allowed to attend to, but rather rely on the attention mask. That said, I haven't tested this code with v.shape[2] > 1 and I think its probably broken elsewhere (likely in the conv cache).